### PR TITLE
Add method to add/remove operations from base worker

### LINKED ISSUE #3363: Add support for Crypten in worker
**Is your feature request related to a problem? Please describe.**
A feature request. Currently, for CrypTen we add all the methods that are "specific" for this framework into the ```BaseWorker``` and we need to have a solution to register/unregister those operations.

**Describe the solution you'd like**
Have separate methods to add/remove support for CrypTen

**Describe alternatives you've considered**
Have dedicated workers, but see the discussion from [here](https://github.com/OpenMined/PySyft/pull/3254)

**Additional context**

### PR DESCRIPTION
**Pull request by gmuraru**
Fixes #3363 

Some initial changes to add the mechanism to add/remove methods from a base worker such that it will be able to support the implemented CrypTen operations.

Currently, there is added only CrypTen support, but I choose to implement it in this way (is only an initial draft) such that we could easily add more frameworks in the future.

WIP: Need to add tests to make sure the add/remove mechanism work as intended.

### CODE REVIEW

File: syft/workers/base.py [hunk 1]
```diff
@@ -1160,6 +1160,18 @@ def _get_msg(self, index):
 
         return sy.serde.deserialize(self.msg_history[index], worker=self)
 
+    def add_crypten_support(self):
+        """Add CrypTen specific methods"""
+        from syft.workers.worker_support import add_support
```
Comment (LaRiffle) [inline syft/workers/base.py:1165]: Probably here the import could be global

File: syft/workers/worker_support.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+from syft.frameworks.crypten.worker_support import add_support_to_worker, remove_support_from_worker
+
+supported_frameworks = {}
+
+if dependency_check.crypten_available:
+    supported_frameworks["crypten"] = [add_support_to_worker, remove_support_from_worker]
+
+
+def add_support(worker: BaseWorker, framework: str) -> None:
+    add_support_to_worker(worker)
```
Comment (LaRiffle) [inline syft/workers/worker_support.py:14]: Use your `supported_frameworks` dict :)

File: syft/frameworks/crypten/worker_support.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    assert rank != None
+
+    return_value = run_party(plan, rank, world_size, master_addr, master_port, (), {})
+    return ObjectMessage(return_value)
+
+
+def add_support_to_worker(worker):
+    worker._message_router[CryptenInit] = types.MethodType(run_crypten_party, worker)
+
+    for method in methods_to_add:
```
Comment (LaRiffle) [inline syft/frameworks/crypten/worker_support.py:53]: maybe don't re-add it if it's already there

Comment (gmuraru) [reply syft/frameworks/crypten/worker_support.py:53]: Done

**Comment by gmuraru** [comment]
> As a suggestion: we may need a function to check if a worker supports a specific framework

Should we keep in each worker a set of supported frameworks? (and have a method with -- ```is_supported(framework_name)``` where we simply check if we support that framework (like we have the methods attached to the worker)

File: test/crypten/test_context.py [hunk 4]
```diff
@@ -1,8 +1,32 @@
...
+
+def test_add_crypten_support(workers):
+    alice = workers["alice"]
+
+    for method in methods_to_add:
+        assert not hasattr(
+            alice, method.__name__
+        ), f"Worker should not have method {method.__name__}"
+
+    alice.add_crypten_support()
```
Comment (iamtrask) [inline test/crypten/test_context.py:19]: What exactly is involved in 'Adding crypten support?" Why would it not be supported by default?

Comment (gmuraru) [reply test/crypten/test_context.py:19]: There was a discussion [here] (https://github.com/OpenMined/PySyft/pull/3254).
Also, added screenshots - I hope I did not miss any information
(http://swarm.cs.pub.ro/~gmuraru/OpenMined/pic1.png)
(http://swarm.cs.pub.ro/~gmuraru/OpenMined/pic2.png)

**TL;DR:** We discussed that maybe the framework information might not be contained in the ```BaseWorker``` and we should have a mechanism to register those "specific framework methods". If this is not wanted, I can switch to something else :D

File: syft/workers/base.py (hunk 1 above)
Comment (gmuraru) [reply syft/workers/base.py:1124]: I left it into the method, because making it global it would create an import error.

```worker_support.py``` needs BaseWorker which is not defined :\

File: pip-dep/requirements.txt [hunk 5]
```diff
@@ -17,3 +17,4 @@ torch~=1.4.0
 tornado==4.5.3
 websocket_client~=0.57.0
 websockets~=8.1.0
+git+https://github.com/gmuraru/CrypTen.git@2aa0c2f14b472a8df6a85be1cfc4630479a10b2f#egg=crypten
```
Comment (gmuraru) [inline pip-dep/requirements.txt:20]: This should be removed after merging the [Fix plan](https://github.com/OpenMined/PySyft/pull/3422) changes

File: syft/frameworks/crypten/context.py [hunk 6]
```diff
@@ -159,6 +160,8 @@ def wrapper(*args, **kwargs):
             if was_initialized:
                 crypten.init()
 
+            sy.local_worker.remove_crypten_support()
```
Comment (LaRiffle) [inline syft/frameworks/crypten/context.py:163]: I wonder if we should remove crypten support?

File: syft/frameworks/crypten/worker_support.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        Args:
+            message (CryptenInit): should contain the rank_to_worker_id, world_size, master_addr and master_port.
+
+        Returns:
+            An ObjectMessage containing the return value of the crypten function computed.
+        """
+
+    worker.rank_to_worker_id, world_size, master_addr, master_port = message.crypten_context
+
+    plans = worker.search("crypten_plan")
```
Comment (LaRiffle) [inline syft/frameworks/crypten/worker_support.py:33]: .search has changed and now no longer returns pointers right?

Comment (LaRiffle) [reply syft/frameworks/crypten/worker_support.py:33]: it's request_search which does

File: syft/frameworks/crypten/worker_support.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            break
+
+    assert rank != None
+
+    return_value = run_party(plan, rank, world_size, master_addr, master_port, (), {})
+    return ObjectMessage(return_value)
+
+
+def add_support_to_worker(worker):
+    worker._message_router[CryptenInit] = types.MethodType(run_crypten_party, worker)
```
Comment (LaRiffle) [inline syft/frameworks/crypten/worker_support.py:51]: What is the role of `types.MethodType` ?

Comment (LaRiffle) [reply syft/frameworks/crypten/worker_support.py:51]: Ah ok :) https://stackoverflow.com/questions/37455426/advantages-of-using-methodtype-in-python

File: syft/workers/worker_support.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+from syft.frameworks.crypten.worker_support import add_support_to_worker, remove_support_from_worker
+
+supported_frameworks = {}
+
+if dependency_check.crypten_available:
+    supported_frameworks["crypten"] = [add_support_to_worker, remove_support_from_worker]
+
+
+def add_support(worker: BaseWorker, framework: str) -> None:
+    supported_frameworks[framework][0](worker)
```
Comment (LaRiffle) [inline syft/workers/worker_support.py:14]: I'm not sure this scales: maybe you could replace the list [add_support_to_worker, remove_support_from_worker] by a dict with "standard" keys, so that if you need to has more on them you don't mess up with indices

File: syft/workers/worker_support.py (hunk 2 above)
Comment (gmuraru) [reply syft/workers/worker_support.py]: Done

File: syft/workers/worker_support.py (hunk 9 above)
Comment (gmuraru) [reply syft/workers/worker_support.py:14]: Done

File: syft/frameworks/crypten/worker_support.py (hunk 7 above)
Comment (gmuraru) [reply syft/frameworks/crypten/worker_support.py:33]: Hmmm...```request_search``` is used if you want to take a plan from the worker that built the plan, right?

In this scenario, the plan is already sent after it was built (to each worker) and the worker will simply search for the plan in the objects they have.

File: pip-dep/requirements.txt [hunk 10]
```diff
@@ -10,7 +10,7 @@ Pillow~=6.2.2
 requests~=2.22.0
 RestrictedPython~=5.0
 scipy~=1.4.1
-syft-proto~=0.3.0
+syft-proto~=0.3.1
```
Comment (youben11) [inline pip-dep/requirements.txt:13]: We can go for the 0.4.1 which got released after this one

File: syft/frameworks/crypten/context.py [hunk 11]
```diff
@@ -87,7 +87,7 @@ def _send_party_info(worker, rank, msg, return_values, model=None):
     return_values[rank] = utils.unpack_values(response.object, model)
 
 
-def run_multiworkers(workers: list, master_addr: str, master_port: int = 15463):
```
Comment (youben11) [inline syft/frameworks/crypten/context.py:90]: Q: I didn't get why you always change ports George :p

File: syft/workers/worker_support.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+from syft.frameworks.crypten.worker_support import add_support_to_worker, remove_support_from_worker
+
+
+supported_frameworks = {}
+
+if dependency_check.crypten_available:
+    supported_frameworks["crypten"] = {
+        "add_support": add_support_to_worker,
+        "remove_support": remove_support_from_worker,
+    }
```
Comment (youben11) [inline syft/workers/worker_support.py:14]: What about adding the framework name in these functions?

```suggestion
from syft.frameworks.crypten.worker_support import add_support_to_worker as add_crypten_support, remove_support_from_worker as remove_crypten_support


supported_frameworks = {}

if dependency_check.crypten_available:
    supported_frameworks["crypten"] = {
        "add_support": add_crypten_support,
        "remove_support": remove_crypten_support,
    }
```

File: syft/frameworks/crypten/context.py (hunk 11 above)
Comment (gmuraru) [reply syft/frameworks/crypten/context.py:90]: Ahh...Because while testing I get the error "port already in use" (if the tests fail...I can change it back)

File: pip-dep/requirements.txt (hunk 10 above)
Comment (gmuraru) [reply pip-dep/requirements.txt:13]: Hmm, first I want to merge this [branch](https://github.com/OpenMined/PySyft/pull/3453) and then this should be fixed :D

File: syft/frameworks/crypten/context.py (hunk 6 above)
Comment (gmuraru) [reply syft/frameworks/crypten/context.py:192]: So you suggest the following:
If we want to run with a plan/jail we add the support to workers and this support should not be removed, until explicitly the user requests that

File: syft/frameworks/crypten/context.py (hunk 11 above)
Comment (gmuraru) [reply syft/frameworks/crypten/context.py:90]: Done

Comment (youben11) [reply syft/frameworks/crypten/context.py]: Ah, that's because you have processes still running in the background. I have a script in my workspace that I call if processes keeps running, it's a single command script, might be helpful:

```bash
#!/bin/bash

kill -9 `ps |grep python | awk '{$1=$1};1' |cut -f1 -d ' '`
```

**Comment by gmuraru** [comment]
This one can be merged -- the coverage is failing -- we need to add more tests :(

**Comment by youben11** [comment]
> This one can be merged -- the coverage is failing -- we need to add more tests :(

That's my fault, I should add full tests for the Jail functionality as soon as I get it to work with training a model.

File: syft/frameworks/crypten/context.py (hunk 6 above)
Comment (gmuraru) [reply syft/frameworks/crypten/context.py:192]: ### TODO: Don't remove this.

File: pip-dep/requirements.txt [hunk 13]
```diff
@@ -10,7 +10,7 @@ Pillow~=6.2.2
 requests~=2.22.0
 RestrictedPython~=5.0
 scipy~=1.4.1
-syft-proto~=0.3.0
+syft-proto~=0.4.1
```
Comment (gmuraru) [inline pip-dep/requirements.txt:13]: Will change this back since the current PR does not have some changes required for tests to be able to pass -- will bump the version in a future PR

File: syft/frameworks/crypten/worker_support.py (hunk 7 above)
Comment (LaRiffle) [reply syft/frameworks/crypten/worker_support.py:36]: All good!

**Comment by gmuraru** [comment]
> As a suggestion: we may need a function to check if a worker supports a specific framework

Ahh...will add this :)
